common/src/main/java/com/graphicsfuzz/common/util/FragmentShaderJobToVertexShaderJob.java

```diff
@@ -0,0 +1,226 @@
+                    new VariableIdentifierExpr("frag_color"),
+                    new VariableIdentifierExpr(Constants.GLF_COLOR),
+                    BinOp.ASSIGN))), false))));
+
+    // Find the first declaration that is not a precision declaration.
+    Declaration firstNonPrecisionDeclaration = null;
+    for (Declaration decl : vertexShader.getTopLevelDeclarations()) {
+      if (decl instanceof PrecisionDeclaration) {
+        continue;
+      }
+      firstNonPrecisionDeclaration = decl;
+      break;
+    }
+
+    assert firstNonPrecisionDeclaration != null;
+
+    // Add a global _GLF_FragCoord that will replace gl_Fragcoord
+    vertexShader.addDeclarationBefore(
+        new VariablesDeclaration(
+            BasicType.VEC4, new VariableDeclInfo(Constants.GLF_FRAGCOORD, null, null)),
+        firstNonPrecisionDeclaration);
+
+    // Add vertex position input
+    vertexShader.addDeclarationBefore(
+        new VariablesDeclaration(new QualifiedType(BasicType.VEC4,
+        Arrays.asList(new LayoutQualifierSequence(
+                new LocationLayoutQualifier(0)),
+            TypeQualifier.SHADER_INPUT)),
+        new VariableDeclInfo(Constants.GLF_POS, null, null)),
+        firstNonPrecisionDeclaration);
+
+    // At start of main, populate _GLF_FragCoord with coordinates calculated from vertex input
+    vertexShader.getMainFunction().getBody().insertStmt(0,
+        new ExprStmt(
+            new BinaryExpr(
+                new VariableIdentifierExpr(Constants.GLF_FRAGCOORD),
+                    new BinaryExpr(
+                    new ParenExpr(
+                    new BinaryExpr(
+                        new VariableIdentifierExpr(Constants.GLF_POS),
+                        new TypeConstructorExpr(BasicType.VEC4.toString(),
+                            new FloatConstantExpr("1.0"),
+                            new FloatConstantExpr("1.0"),
+                            new FloatConstantExpr("0.0"),
+                            new FloatConstantExpr("0.0")),
+                        BinOp.ADD)),
+                        new TypeConstructorExpr(BasicType.VEC4.toString(),
+                            new FloatConstantExpr("128.0"),
+                            new FloatConstantExpr("128.0"),
+                            new FloatConstantExpr("1.0"),
+                            new FloatConstantExpr("1.0")),
+                        BinOp.MUL),
+            BinOp.ASSIGN)));
+
+    // Replace all instances of gl_FragCoord with _GLF_FragCoord
+    new StandardVisitor() {
+      @Override
+      public void visitVariableIdentifierExpr(VariableIdentifierExpr variableIdentifierExpr) {
+        super.visitVariableIdentifierExpr(variableIdentifierExpr);
+        if (variableIdentifierExpr.getName().equals(OpenGlConstants.GL_FRAG_COORD)) {
+          variableIdentifierExpr.setName(Constants.GLF_FRAGCOORD);
+        }
+      }
+    }.visit(vertexShader);
+
+    // Add both shaders to list
+    shaders.add(fragmentShader);
```

Comment:
You could declare 'shaders' at this point.

You could even eliminate 'shaders' and pass Arrays.asList(fragmentShader, vertexShader) instead of the 'shaders' argument to the 'GlslShaderJob' constructor.

Nit: I tend to prefer to put vertex shaders first (to reflect the graphics pipeline), but it doesn't really matter.

Reply:
Eliminated and reordered.